### ISSUE TITLE
Return 503 if route is registered in consul, but no healthy ainstances available

### ISSUE THREAD

**Issue by mantasaudickas** [fabiolb/fabio#511]
Hello,

together with fabio we build our url's hierarchically, for example:
/ -> root application (some kind of dashboard)
/api -> dashboard API
/sales -> sales application UI
/sales/api -> sales application API

So lets say for some reason we deploy wrong version to sales API service, which fails to start (thus consul marks it as critical and fabio removes from routing).

Now if another service calls /sales/api - request will land in sales UI.. which has nothing to do with that request.

So my question - is it possible to keep /sales/api route even if service is not available?
I know about this variable: https://fabiolb.net/ref/registry.consul.service.status/
So, yes, I can specify that critical services should be kept, but if I have load balanced environment with one service failing, part of my requests will fail.

So I would like to have a functionality, which identifies valid route but which does not have any working services now - it would result in 503 (service is not available) instead of forwarding request to UI.

Would it be possible somehow?

**Comment by leprechau** [comment]
The behavior you described is correct.  I'm concerned that any feature that allowed keeping a route that was no longer present in the backend could lead to a stale routing table which would also cause problems.

The easiest "fix" here would be to have your UI container present a "pretty" 404 page.  I don't see how Fabio could ever know that `/sales/api` is a valid route that should *not* match `/sales` but `/sales/css/something.css` should map to the `/sales` route.

**Comment by leprechau** [comment]
Sorry, I just read your issue again and I think I understand your actual request.  You basically want to define the behavior for "unhealthy" prefix matches?

What would you think of some sort of maintenance page feature?

Basic idea... if a incoming request would match a route but that service is either failed or marked as maintenance you could specify a custom response.  This would be similar to how the `noroutestatus` and `noroutehtmlpath` options currently work.

**Comment by mantasaudickas** [comment]
Well.. Consul is Fabio database.. so it just needs to use it.
In my case services with `urlprefix-/sales/api` tags are already in Consul, but lets say for some reason all these instances are failing and /status endpoint is returning 503.
In that case Fabio is simply `forgetting` this route and searching another match, while practically - service route is known, just not functional.
So in that case responding to client with `noroutestatus` or 'noroutehtmlpath` would be fine as long as I don't have to specify anything manually for each service (we have many of them).

**Comment by leprechau** [comment]
Currently `no route` means no route match in the table.  That's not what is happening here.  There is a route matching the request in the table.  Fabio is currently doing the *right* thing.  This would be new functionality that would be similar but different to the current `noroute*` options.

**Comment by mantasaudickas** [comment]
Yes, you are right.

**Comment by leprechau** [comment]
Okay, I understand what you're asking for now and I can see the use case.  Would you be able to submit a PR?

**Comment by mantasaudickas** [comment]
Never tried to code in GO.. might take a while :)

**Comment by leprechau** [comment]
No problem, I'll try and take a look at it in the next few days.

**Comment by kuskmen** [comment]
@leprechau I've always wanted to join fabio project and this seems like a suitable case to do. Do you mind if I try to prepare a PR in next few days? :)

**Comment by leprechau** [comment]
Absolutely, that would be great!

**Comment by kuskmen** [comment]
Hi guys, I looked upon the implementation in fabio and I was wondering if you can provide some more detailed description/requirement of what exactly is needed/desired.

> In my case services with urlprefix-/sales/api tags are already in Consul, but lets say for some reason all these instances are failing and /status endpoint is returning 503. In that case Fabio is simply forgetting this route and searching another match, while practically - service route is known, just not functional.

As far as I saw these routes are indeed "forgotten" (unless I didn't see something in the implementation) so I was wondering are we saying that we should keep those routes even if the health check is critical and respond with 503 when such routes are requested or I completely misunderstood the idea?

**Comment by mantasaudickas** [comment]
Yes, that's correct. Routes are available in Consul, but if all registered instances are down (warning and/or critical state) - fabio should respond with 503 instead of selecting another matching route.
Which consul states are unroutable should be decided using 'registry.consul.service.status' variable.

**Comment by kuskmen** [comment]
@leprechau , @magiconair  do you agree with the terms guys?

**Comment by leprechau** [comment]
@kuskmen @mantasaudickas I see the reasoning but I'm a bit concerned about adding more state to the current routing table.

**Comment by sourcekick** [comment]
> Return 503 if route is registered in consul, but no healthy instances are available

I agree with that.
